[Unity] Merge with upstream

jwfromm opened the pull request:
We need to grab a few changes to enable the scorecard.

hypercubestart commented on python/tvm/relax/block_builder.py:
```diff
@@ -259,22 +262,22 @@ def dataflow(self) -> DataflowScope:
-        """Emit a call node.
-        This infers the shape and type of the CallNode, create a variable,
-        and bind the CallNode to the variable.
+    def emit(self, expr: relay.Expr) -> Var:
```
i think its better to import Expr from relax namespace

hypercubestart commented on python/tvm/relax/block_builder.py:
```diff
@@ -548,3 +548,32 @@ def get_unique_name(self, name_prefix: str) -> str:
+        gvar : GlobalVar
+            The global var bound to the tir_func.
+        """
+        prim_func_hash = tvm.ir.structural_hash(tir_func)
```
do we need to worry about hash collision? 

also dumb question: why does this work if the `map_free_vars` argument to `structural_hash` is False? Two separately generated tir_func should have different memory locations for variables

YuchenJin replied:
Good points! The comment of `structural_hash` says collision is rare but we probably should avoid the collision using `structural_equal` if it's inevitable. I will update the code.

Now I feel we should set `map_free_vars` to True, the test works just because it does not contain free variables. :)

hypercubestart replied:
oh haha i missed the free variables part, thanks I agree it should be set to true

YuchenJin replied:
@hypercubestart, I move the dedup logic to C++ to avoid hash collisions. Could you take a second look? Thanks!

tqchen commented on include/tvm/relax/block_builder.h:
```diff
@@ -141,6 +141,22 @@ class BlockBuilderNode : public Object {
+   * GlobalVar directly.
+   * \return The global var bound to the added function.
+   */
+  GlobalVar AddFunc(const BaseFunc& func, const std::string& func_name);
```
prefer String

YuchenJin replied:
Yes, it is better! Thanks for the suggestions!

hypercubestart reviewed (approved):
lgtm thanks!

psrivas2 committed:
[Unity][Fix] Allow scalar layout initialization (#14370)

Fix assert to allow scalar (ndim=0) layout initialization.

areusch committed:
[Unity] Also include output dtype in simt MathInstruction (#14372)

Also include output dtype in simt MathInstruction.

vinx13 committed:
[Unity][VM] Add CUDA graph vm builtins (#14371)

This PR added Relax VM builtin functions to execute with CUDA graph.

- vm.builtin.cuda_graph.get_cached_alloc: Allocate and cache storage objects for future vm invocation
- vm.builtin.cuda_graph.run_or_capture: Launched captured CUDA graph or capture the CUDA graph using CUDA API and save in the cache
The graph rewriting to enable CUDA graph backend will be done in a separate PR.

Commit:
[Unity] Add missing #include <array> (#14383)

The file tests/cpp/nested_msg_test.cc may fail to compile if <array>
is not included explicitly.

spectrometerHBH committed:
[Unity][Transform] SplitCallTIRByPattern and CUTLASS backend (#14274)

Currently, the BYOC system is based on op-level pattern matching, this PR intends to provide primary support for TIR-level pattern matching based on backend registration and dispatching.

For now, it simply matches the first set of for loops in PrimFunc.

Co-authored-by: Hongyi Jin (@jinhongyii)

Commit:
[Unity] Support simple dynamic-shape-aware fusion (#14396)

This PR adds support for simple dynamic-shape-aware fusion, which is the first step towards supporting dynamic shapes. The main changes are as follows:

- Fix FuncStructInfo in well-formed checks
- Renew symbolic var defs in fuse_ops to prevent malformed functions

Commit:
Merge remote-tracking branch 'tvm/unity' into unity_merge
